modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java

```diff
@@ -4190,4 +4224,150 @@ public boolean addTimeoutHandler() {
             "mappings", mappings,
             "super", super.toString());
     }
+
+    /**
+     * Creates savepoint.
+     *
+     * @param name Savepoint ID.
+     * @param overwrite Overwrite savepoint flag.
+     */
+    @Override public void savepoint(String name, boolean overwrite) throws IgniteCheckedException {
```

Comment:
probably, your api should be thread safe

Reply:
This is not multithreaded code. Suspend() and resume() are syncronized, so any savepoint created before suspend() is visible after resume().

Reply:
I meant org.apache.ignite.internal.processors.cache.transactions.TransactionProxyImpl#savepoint(java.lang.String, boolean) should be thread safe

Reply:
Why? We can't call this method simultaneously from different threads, and there is synchronization between threads in suspend/resume.

Reply:
...